src/power_grid_model/core/power_grid_dataset.py

```diff
@@ -266,6 +274,23 @@ def _validate_properties(self, data: ComponentData, schema: ComponentMetaData):
         if properties.batch_size != self._batch_size:
             raise ValueError(f"Dataset must have a consistent batch size across all components. {VALIDATOR_MSG}")
 
+    def _validate_dtypes_compatibility(self, data: ComponentData, schema: ComponentMetaData):
+        sub_data = _extract_data_from_component_data(data)
+        if is_columnar(data):
+            for attr, array in sub_data.items():
+                if (
+                    schema.dtype.names is None
+                    or attr not in schema.dtype.names
+                    or (schema.dtype[attr].shape == (3,) and array.shape[-1] != 3)
+                ):
+                    raise TypeError("Given data has a different schema than supported.")
+                if array.dtype.base != schema.dtype[attr].base:
+                    warnings.warn(
```

Comment:
Do we need such a complicated loop? We can just check it in the data view right?

We check before the line:

https://github.com/PowerGridModel/power-grid-model/blob/f90cf37da600762552b6b028e5bf5cf1beb5aab0/src/power_grid_model/core/buffer_handling.py#L86

By

```python
if data.dtype != dtype:
    warning.warn(...)
```